proposals/handler.go

```diff
@@ -0,0 +1,380 @@
+package proposals
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/log"
+	"github.com/spacemeshos/go-spacemesh/p2p/pubsub"
+	"github.com/spacemeshos/go-spacemesh/proposals/metrics"
+	"github.com/spacemeshos/go-spacemesh/system"
+)
+
+// NewBallotProtocol is the protocol indicator for gossip Ballots.
+const NewBallotProtocol = "newBallot"
+
+// NewProposalProtocol is the protocol indicator for gossip Proposals.
+const NewProposalProtocol = "newProposal"
+
+var (
+	errMalformedData         = errors.New("malformed data")
+	errInitialize            = errors.New("failed to initialize")
+	errInvalidATXID          = errors.New("ballot has invalid ATXID")
+	errMissingBaseBallot     = errors.New("base ballot is missing")
+	errMissingEpochData      = errors.New("epoch data is missing in ref ballot")
+	errUnexpectedEpochData   = errors.New("non-ref ballot declares epoch data")
+	errEmptyActiveSet        = errors.New("ref ballot declares empty active set")
+	errMissingBeacon         = errors.New("beacon is missing in ref ballot")
+	errNotEligible           = errors.New("ballot not eligible")
+	errConflictingExceptions = errors.New("conflicting exceptions")
+	errExceptionsOverflow    = errors.New("too many exceptions")
+	errDuplicateTX           = errors.New("duplicate TxID in proposal")
+	errDuplicateATX          = errors.New("duplicate ATXID in active set")
+)
+
+// Handler processes Proposal from gossip and, if deems it valid, propagates it to peers.
+type Handler struct {
+	logger log.Log
+	cfg    Config
+
+	fetcher   system.Fetcher
+	mesh      mesh
+	validator eligibilityValidator
+}
+
+// Config defines configuration for the handler.
+type Config struct {
+	LayerSize            uint32
+	LayersPerEpoch       uint32
+	GoldenATXID          types.ATXID
+	MaxExceptions        int
+	EligibilityValidator eligibilityValidator
+}
+
+// defaultConfig for BlockHandler.
+func defaultConfig() Config {
+	return Config{
+		MaxExceptions:        1000,
+		EligibilityValidator: nil,
+	}
+}
+
+// Opt for configuring BlockHandler.
+type Opt func(h *Handler)
+
+// WithGoldenATXID defines the golden ATXID.
+func WithGoldenATXID(atx types.ATXID) Opt {
+	return func(h *Handler) {
+		h.cfg.GoldenATXID = atx
+	}
+}
+
+// WithLayerSize defines the average number of proposal per layer.
+func WithLayerSize(size uint32) Opt {
+	return func(h *Handler) {
+		h.cfg.LayerSize = size
+	}
+}
+
+// WithLayerPerEpoch defines the number of layers per epoch.
+func WithLayerPerEpoch(layers uint32) Opt {
+	return func(h *Handler) {
+		h.cfg.LayersPerEpoch = layers
+	}
+}
+
+// WithValidator defines eligibility validator.
+func WithValidator(v eligibilityValidator) Opt {
```

Comment:
is it only for testing? can use private option if so

Reply:
done